addons/binding/org.openhab.binding.nibeuplink/ESH-INF/thing/base-channel-types.xml

```diff
@@ -0,0 +1,209 @@
+		<description>Total electric additive operation time</description>
+		<state pattern="%.1f h" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-43084">
+		<item-type>Number:Power</item-type>
+		<label>Int. el.add. Power</label>
+		<description>Current power from the internal electrical addition</description>
+		<state pattern="%.2f kW" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-43161">
+		<item-type>Number</item-type>
+		<label>External adjustment activated via input S1</label>
+		<description></description>
+		<state pattern="%d" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-44298">
+		<item-type>Number:Energy</item-type>
+		<label>Heat Meter - HW Cpr and Add EP14</label>
+		<description>Accumulated energy production as calculated by the heat meter</description>
+		<state pattern="%.1f kWh" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-44300">
+		<item-type>Number:Energy</item-type>
+		<label>Heat Meter - Heat Cpr and Add EP14</label>
+		<description>Accumulated energy production as calculated by the heat meter</description>
+		<state pattern="%.1f kWh" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-44304">
+		<item-type>Number:Energy</item-type>
+		<label>Heat Meter - Pool Cpr EP14</label>
+		<description>Accumulated energy production as calculated by the heat meter</description>
+		<state pattern="%.1f kWh" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-44306">
+		<item-type>Number:Energy</item-type>
+		<label>Heat Meter - HW Cpr EP14</label>
+		<description>Accumulated energy production as calculated by the heat meter</description>
+		<state pattern="%.1f kWh" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-44308">
+		<item-type>Number:Energy</item-type>
+		<label>Heat Meter - Heat Cpr EP14</label>
+		<description>Accumulated energy production as calculated by the heat meter</description>
+		<state pattern="%.1f kWh" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-47041">
+		<item-type>Number</item-type>
+		<label>Hot water mode</label>
+		<description> 0=Economy 1=Normal 2=Luxury</description>
+		<state pattern="%d" readOnly="false">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-47212">
+		<item-type>Number:Power</item-type>
+		<label>Max int add. power</label>
+		<description></description>
+		<state min="0" max="4500" step="1" pattern="%.2f kW" readOnly="false">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-48043">
+		<item-type>Number</item-type>
+		<label>Holiday</label>
+		<description>0=inactive, 10=active</description>
+		<state pattern="%d" readOnly="false">
+		</state>
+	</channel-type>
+	<channel-type id="base-type-48132">
+		<item-type>Number</item-type>
+		<label>Temporary Lux</label>
+		<description>0=Off, 1=3h, 2=6h, 3=12h, 4=One time increase</description>
```

Comment:
Could you please provide an example or link to further docs?
I had already tried to map values like this:

```
	<channel-type id="vvm320-type-48132">
		<item-type>String</item-type>
		<label>Temporary Lux</label>
		<description>0=Off, 1=3h, 2=6h, 3=12h, 4=One time increase</description>
		<state pattern="%s" readOnly="false">
			<options>
				<option value="0">Off</option>
				<option value="1">3h</option>
				<option value="2">6h</option>
				<option value="3">12h</option>
				<option value="4">One time</option>
			</options>
		</state>
	</channel-type>
```
However this is incompatible with the type number, at least it did not work. Also I did not relaize how to translate it. The current implementation using the raw number values has 2 advantages

1. values can be stored in influxdb+grafana (as numbers are supported, strings do not make sense)
2. user can use user defined mappings to translate number into any string.

Maybe this is still not the best way to solve it but at least it is better than the string stuff I did before.

Reply:
> this is incompatible with the type number

The binding is responsible to map the numbers form the device to strings for the channel and vice versa.

> [number] values can be stored in influxdb+grafana

true, also I would assume discrete values would also work.

> user defined mappings to translate number into any string

Can be used nonetheless, even for fixed option. Mapping is then string <-> string.

The code snippet looks good, I would only suggest to change the description to something describing the field better. The options then document themselves. 
Translation for state options is described here: https://www.eclipse.org/smarthome/documentation/features/internationalization.html#thing-types

Reply:
I just checked the snippet in a demo setup and it worked just fine. I did some small changes though:

* remove the `pattern` attribute
* set the values into quotes

result:

```
<channel-type id="vvm320-type-48132">
    <item-type>Number</item-type>
    <label>Temporary Lux</label>
    <state readOnly="false">
        <options>
            <option value="0">Off</option>
            <option value="1">3h</option>
            <option value="2">6h</option>
            <option value="3">12h</option>
            <option value="4">One time</option>
        </options>
    </state>
</channel-type>
```

Reply:
Well in general this works now. But unfortunately this approach suffers from the same strange behaviour like the other translations via mapping files. Depending on where you want to show the values you need different mappings.

```
	<channel-type id="vvm320-type-47260">
		<item-type>Number</item-type>
		<label>Selected Fan speed</label>
		<description></description>
		<state readOnly="false">
			<options>
                <option value="0">normal 0</option>
                <option value="0.0">normal 0.0</option>
                <option value="1">speed 1</option>
                <option value="1.0">speed 1.0</option>
                <option value="2">speed 2</option>
                <option value="2.0">speed 2.0</option>
                <option value="3">speed 3</option>
                <option value="3.0">speed 3.0</option>
                <option value="4">speed 4</option>
                <option value="4.0">speed 4.0</option>
			</options>
		</state>
	</channel-type>
```
The basic UI renders numbers as numbers in general, no mapping is applied. If you specify a selection in a sitemap mapping is applied automatically but you will need decimals (0.0, 1.0, ...). In paperui it only renders correctly when using integer mappings (0, 1, 2, ...)

I do not like this because all options are shown in both UIs so you will have duplicates.

Reply:
I agree, those duplicates are very ugly. How comes that sometimes integer and sometimes decimals are used? I think the binding should take care that only integers are exposed to the channels regardless of what the device delivers. Are you able to make a conversion in the binding so that decimals will be turned to ints?

Reply:
I do not know why BasicUI has a different behaviour than PaperUI. But at least I could fix it in my latest commit. I did a conversion to double. That was necessary in the past when I used another API. But this RAW API always returns integer/long values so no double conversion is needed. I removed that stuff and now it works fine.
I wonder why this really makes the difference becaus I used NumberType before and this part did not change. So NumberType seems to have a behaviour depending on the value which was put in on construction.

As I wrote before: In BasicUI mapping is only applied in sitemaps when I use sitemap item type "Selection" but if I just dig into my hierarchical group representation each item is displayed as its type that comes from the .items definition file. So it is displayed as number unless I apply a value mapper in the .items definition. I guess there is no way to use the state mappings?